From 360080a1244dc806af0773fecf0802846c2354dd Mon Sep 17 00:00:00 2001
From: Paolo Bonzini <pbonzini@redhat.com>
Date: Tue, 22 Jul 2014 14:38:50 -0400
Subject: [virt] kvm/x86: mark hyper-v vapic assist page as dirty

Message-id: <1406040016-3289-136-git-send-email-pbonzini@redhat.com>
Patchwork-id: 86214
O-Subject: [RHEL7 PATCH v2 135/221] KVM: x86: mark hyper-v vapic assist page as dirty
Bugzilla: 1116936
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Bandan Das <bsd@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Vadim Rozenfeld <vrozenfe@redhat.com>

Signed-off-by: Vadim Rozenfeld <vrozenfe@redhat.com>
Reviewed-by: Marcelo Tosatti <mtosatti@redhat.com>
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
(cherry picked from commit b3af1e889ec4909f6b48dabd19a311d9c9f8d58e)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index 86f9c8d..b085d21 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -1902,19 +1902,21 @@ static int set_msr_hyperv(struct kvm_vcpu *vcpu, u32 msr, u64 data)
 {
  switch (msr) {
  case HV_X64_MSR_APIC_ASSIST_PAGE: {
+  u64 gfn;
   unsigned long addr;
 
   if (!(data & HV_X64_MSR_APIC_ASSIST_PAGE_ENABLE)) {
    vcpu->arch.hv_vapic = data;
    break;
   }
-  addr = gfn_to_hva(vcpu->kvm, data >>
-      HV_X64_MSR_APIC_ASSIST_PAGE_ADDRESS_SHIFT);
+  gfn = data >> HV_X64_MSR_APIC_ASSIST_PAGE_ADDRESS_SHIFT;
+  addr = gfn_to_hva(vcpu->kvm, gfn);
   if (kvm_is_error_hva(addr))
    return 1;
   if (__clear_user((void __user *)addr, PAGE_SIZE))
    return 1;
   vcpu->arch.hv_vapic = data;
+  mark_page_dirty(vcpu->kvm, gfn);
   break;
  }
  case HV_X64_MSR_EOI:
-- 
1.7.1